### ISSUE TITLE
Shareable directive doesn't behave as expected with extended types

### ISSUE THREAD

**Issue by gwardwell** [apollographql/federation#2135]
## Description

When the same subgraph has a `@shareable` value type that is then extended, Rover does not compile the type in an expected way.

## Expected behavior

When I author schema like the following:

```gql
# subgraph-a.graphqls
type Example @shareable {
  foo: String!
}

extend type Example {
  bar: String!
}
```

```gql
# subgraph-b.graphqls
type Example @shareable {
  foo: String!
  bar: String!
}
```

I would expect Rover to successfully generate the following supergraph schema (minus decoration):

```gql
# Supergraph Schema
type Example @shareable {
  foo: String!
  bar: String!
}
```

## Actual behavior

The following schema results in the following Rover error:

```gql
# subgraph-a.graphqls
type Example @shareable {
  foo: String!
}

extend type Example {
  bar: String!
}
```

```gql
# subgraph-b.graphqls
type Example @shareable {
  foo: String!
  bar: String!
}
```

```
composing supergraph with Federation v2.1.1.
error[E029]: Encountered 1 build error while trying to build a supergraph.

Caused by:
    UNKNOWN: Non-shareable field "Example.bar" is resolved from multiple subgraphs: it is resolved from subgraphs "subgraph_a" and "subgraph_b" and defined as non-shareable in subgraph "subgraph_a"
    
        The subgraph schemas you provided are incompatible with each other. See https://www.apollographql.com/docs/federation/errors/ for more information on resolving build errors.
```

If I add `@shareable` to the extended `Example` type, I get the following error:

```gql
# subgraph-a.graphqls
type Example @shareable {
  foo: String!
}

extend type Example @shareable {
  bar: String!
}
```

```gql
# subgraph-b.graphqls
type Example @shareable {
  foo: String!
  bar: String!
}
```

```
composing supergraph with Federation v2.1.1.
error[E029]: Encountered 1 build error while trying to build a supergraph.

Caused by:
    UNKNOWN: [subgraph_a] The directive "@shareable" can only be used once at this location.
    
        The subgraph schemas you provided are incompatible with each other. See https://www.apollographql.com/docs/federation/errors/ for more information on resolving build errors.
```

Only by marking the individual fields as `@shareable` in the extended type will successfully compile:

```gql
# subgraph-a.graphqls
type Example @shareable {
  foo: String!
}

extend type Example {
  bar: String! @shareable
}
```

```gql
# subgraph-b.graphqls
type Example @shareable {
  foo: String!
  bar: String!
}
```

## Reproduction

Any of the above `subgraph-a.graphqls` and `subgraph-b.graphqls` can be used in conjunction with the the following `supergraph.yaml` file and Rover command:

```yaml
# supergraph.yaml
federation_version: 2
subgraphs:
  subgraph_a:
    routing_url: https://films.example.com
    schema:
      file: ./subgraph-a.graphqls
  subgraph_b:
    routing_url: https://films.example.com
    schema:
      file: ./subgraph-b.graphqls
```

```
rover supergraph compose --config ./supergraph.yaml > supergraph.graphql
```

**Comment by pcmanus** [comment]
The fact that `@shareable` applied on a type definition only applies to the field within that particular definition is kind of intentional. The rough rationales for that are that:
- since `@shareable` is really more of a field property (and applying on type is more of a convenience), we felt that having it apply to fields that may be declared in an extension potentially "far away" (maybe in a separate file?) could be hard to follow/read and maybe surprising.
- this also gives a syntactical way to segregate fields between shareable one and others if you wanted to, which we felt may prove useful at times.
Not to suggest that alternatives wouldn't have other advantages, but well, it's the choice we made. 

But the fact that this currently mean that you have resort to field-based `@shareable` when you _do_ have extensions of a type for some reason and want everything `@shareable` is a bit more of a oversight to be honest. And I agree it's rather inconvenient.

Anyway, I need to sleep on it to make sure I'm not missing some obvious downside, but I think the obvious improvements here would be to make `@shareable` repeatable (exactly for that purpose). So far, the only minor downside I can think of is that it may not be immediately obvious to new user why its repeatable given it has no arguments, but that's rather minor (and nothing that cannot be fixed by good documentation).

I guess one minor question if we go that route might be whether for something like:
```graphql
type T @shareable @shareable {
  ...
}
```
we want to:
1. reject it.
2. let it be and simply have it being redundant/useless.

No particular strong opinion on that on my hand. I tend to personally favor more nitpicky software so I'd probably go for option 1 but a fairly subjective opinion so can easily be convinced with option 2.

**Comment by gwardwell** [comment]
If it changes anything, we plan to introduce linting rules that enforce `@shareable` consistency of all fields on a type to avoid unresolvable fields. Currently we're planning on doing this by enforcing `@shareable` on types rather than fields. We found too many instances where allowing it on fields allowed divergent types an errors, and couldn't come up with a reason why `@shareable` on fields made sense (I'm open to hearing reasons why if anyone has them, FWIW).

In that case, I would _hope_ that either:
1. extended types wouldn't need the directive _or_ 
2. extended types can have the directive

If neither of those can happen, then I'd probably choose to do one of the following:

1. Flip my rule and not allow the use of `@shareable` on types and instead require it on fields, but that seems super inconvenient
2. Decorate the schema myself when publishing and translate type-level `@shareable` directives to the fields

**Comment by samjcombs** [comment]
RE:
> ```graphql
> type T @shareable @shareable {
>   ...
> }
> ```
> 
> we want to:
> 
> 1. reject it.
> 2. let it be and simply have it being redundant/useless.

I vote for #1, rejecting it because it isn't the intended use of the directive, and possibly at that time use the error message to educate as to the intended repeatable use case, as well as pointing to docs where possibly some of the above can be discussed.

Good info above, thanks!

**Comment by gwardwell** [comment]
I can get behind this being an error:

```gql
type T @shareable @shareable {
  ...
}
```

as long as this is allowed:

```gql
type T @shareable {
  ...
}

extend type T @shareable {
  ...
}
```

_or this_:

```gql
type T @shareable {
  ...
}

extend type T { # Inherits @shareable from the non-extension type
  ...
}
```

**Comment by pcmanus** [comment]
I'll note that it is part of the design (and implementation) of `@shareable` that a type may mix shareable and non-shareable. And it's actually commonly the case for entity types. Or rather, while most entity types probably only have non-shareable fields, you have to mark some field `@shareable` if you want to do some `@provides`-like optimisations. But making all fields of an entity type shareable tends to be a bad idea, so it end up being a mix in those cases. I also think that a handful of `@shareable` field in a root type can have some use (but obviously, most field still won't be shareable).

It's true that for non-entity-non-root types, there is currently little reason to mix shareable and non-shareable fields. But even for those, that's more a consequence of the relative inflexibility federation currently offers around distributing non-entity types accross subgraph, and hopefully that will evolve.

Overall, `@shareable` is intrinsically a field property, not a type one. It's main goal is to provide an explicit/direct (but also verified) way to know, if a particular field of a particular subgraph may or may not be provided by other subgraphs, without requiring knowledge of those other subgraphs. It's certainly common that fields of the same type will tend to have the same "shareability", but there is no reason for that to be always true.

All this to say that what I'm pushing for is neither of your 2 alternatives above, @gwardwell, at least not exactly. What I'm pushing for is a relaxed version of your first alternative. Essentially:
```graphql
type T @shareable {
  a: Int
}

# Fine and allowed. It just means `b` is not `@shareable` even if `a` is.
extend type T {
  b: Int
}

# Also fine and allowed. Means `c` is `@shareable`.
extend type T @shareable {
  c: Int
}
```

Now, it's worth discussing the 2nd alternative you propose above: to have type extensions "silently" inherit the `@shareable` of their corresponding type definitions.

I don't like that alternative to be honest, mostly because I think it is at odds with the `@shareable` design goals, but also because it would kind of be a breaking change. Let me expand:
1. One of the main goal of `@shareable` is to make it easy for someone reading a schema to know if a particular field is guaranteed to be "exclusive" to this subgraph or not. That works much better, imo, if `@shareable` is syntactically close to the fields to which it applies. That's really the impetus for the current implementation only considering `@shareable` on the very definition in which a field is defined. But if, for type extensions, you need to find the corresponding type definition (which is likely not "close", probably not even the same file; why use an extension in the first place otherwise?), then that weakens the proposition somewhat. Feels like a step backward.
2. As said, having types mix shareable and non-shareable does happen (and there may even be more reason to do in the future). And I think some user may well found values is using type extensions as a way to conveniently segregate fields that should be `@shareable` from other fields (think an entity type with tons of fields). This is more of a nice-to-have for a specific cases, certainly, but it's nice to have nice-to-have, no?
3. This would technically be a breaking change. Meaning, it shouldn't break composition for anyone (at least I don't think so), but anyone relying on the guarantees of the current behaviour may not realise that those guarantees are gone. I'd certainly prefer avoiding it.

**Comment by gwardwell** [comment]
@pcmanus Do you have examples of `@shareable` being necessary on an entity? So far, schema checks have indicated to me that `@shareable` is not necessary for entities and that entity fields behave as if they are shareable by default.

Your proposed behavior is _exactly_ what I would be looking for (it aligns with exactly what I was thinking for my first option above, I just didn't elaborate quite as well as you did 😁 ). I would love if `extend type T` would require `@shareable` in order for its fields to be shareable. That would provide consistency of behavior between `type T` and `extend type T`, which I think would result in a good dev experience.

The second option was not my ideal scenario, so I am happy to not have `@shareable` silently inherited. I am more than happy to strike that one from the possible outcomes here.

**Comment by pcmanus** [comment]
> entity fields behave as if they are shareable by default

That shouldn't be the case, though one exception is entity key fields, which _are_ shareable by default (the reasoning here is that a non-shareable key field doesn't really make much sense so we decided that exception was probably worth it), so maybe that's what give that impression.

But otherwise, for a field to be shareable means that the field can be resolved by more than one subgraph (I'll note here that because this is about "resolvability", `@external` fields don't really count). Or put another way, a non-shareable field cannot be declared in such a way that it would be resolved by more than one subgraph. And so if you have:
```graphql
# Subgraph A
type User @key(fields: "id") {
  id: ID!
  name: String
  email: String
}
```
```graphql
# Subgraph B
type User @key(fields: "id") {
  id: ID!
  name: String
  age: Int
}
```
then this won't compose, because `name` (a non-key entity field) would have to be resolved by both subgraphs, but it isn't shareable by default (again, `id` is fine since as said above, it's a key field and those are the one exception).

> Do you have examples of @shareable being necessary on an entity?

Well, depends on what we mean by "necessary" I suppose :).

Using `@shareable` on an entity field is only ever useful if you want to do some "duplication" optimisation. In that sense, it's never truly "necessary" but it can be "desirable".

Another way to put it is that  in an entity, the router/gateway can always jump to a different subgraph if a field is not resolved by a given subgraph, so you never "need" multiple subgraphs to resolve a field (aka, having it shareable), but having multiple subgraph resolve the field can save having to jump sometimes.

Long story short, if you do want the `name` field in the example above to be resolved by both subgraph, then you need to mark it `@shareable`. But most entity field are not marked `@shareable`, because resolving an entity field from multiple subgraph is not often doable or desirable (but again, it can make query plans more efficient so it has some specific uses).

I'll lastly note that a special case of such optimisation is actually the use of `@provides`, which allows a subgraph to declare that it can resolve a field, but only on some specific "paths". If you use `@provides` on an entity field, then you _must_ mark that entity field `@shareable` in the other subgraph(s) that resolve it.
